### PR TITLE
Testing branch

### PR DESCRIPTION
**Pull request by hamzash10**
part-2 testing

### CODE REVIEW

File: game_tests.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+
+class unit_tests(unittest.TestCase):
+
+    def setUp(self):
+        self.test_game_state = chess_engine.game_state()
+        self.test_game_state.board = [[Player.EMPTY for _ in range(8)] for _ in range(8)]
+
+    def test_knight_valid_piece_takes_1(self):
+        with patch.object(chess_engine.game_state, 'is_valid_piece', return_value=False):
```
Comment (excellenteam-scaleup) [inline game_tests.py]: missing doc

File: game_tests.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
+import unittest
+from unittest.mock import patch, Mock
+
+import chess_engine
+import Piece
+from enums import Player
+from ai_engine import chess_ai
+
+
```
Comment (excellenteam-scaleup) [inline game_tests.py]: Consider splitting the tests into different files

File: game_tests.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+from ai_engine import chess_ai
+
+
+class unit_tests(unittest.TestCase):
+
+    def setUp(self):
+        self.test_game_state = chess_engine.game_state()
+        self.test_game_state.board = [[Player.EMPTY for _ in range(8)] for _ in range(8)]
+
+    def test_knight_valid_piece_takes_1(self):
```
Comment (excellenteam-scaleup) [inline game_tests.py]: Good division to 3 tests for the two functions.
Choose a more indicative name, replace "...takes_1" to "takes corner/center" accordingly, or use documentation to be more clear.

File: game_tests.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        board = self.test_game_state
+        knight = chess_engine.Knight('n', 3, 4, Player.PLAYER_1)
+        pawn = chess_engine.Pawn('p', 5, 3, Player.PLAYER_2)
+        board.board[3][4] = knight
+        board.board[5][3] = pawn
+        expected_takes = {(5, 5), (4, 6), (4, 2), (2, 2), (2, 6), (1, 3), (1, 5)}
+        valid_takes = set(knight.get_valid_peaceful_moves(board))
+        self.assertEqual(expected_takes, valid_takes)
+
+    def set_full_takes(self):
```
Comment (excellenteam-scaleup) [inline game_tests.py]: It's good that you've separated a logic you're using into a different function, I'd just place it better, perhaps above all the tests. This way the code will "flow" more aesthetically : c;ass definition, functions, then all of the tests. Now it's test1-test1-test2-test2-function-test3-test3.

Comment (excellenteam-scaleup) [reply game_tests.py]: Also, if you see that this functionality is to be used only internally you can consider making it an internal function (_set_full_takes)

File: game_tests.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+        pawn1 = chess_engine.Pawn('p', 1, 3, Player.PLAYER_2)
+        pawn2 = chess_engine.Pawn('p', 5, 5, Player.PLAYER_2)
+        self.test_game_state.board[1][3] = pawn1
+        self.test_game_state.board[5][5] = pawn2
+        expected_peaceful_moves = [(1, 5), (2, 2), (2, 6), (4, 2), (4, 6), (5, 3)]
+        expected_takes = [(1, 3), (5, 5)]
+        expected_moves = expected_peaceful_moves + expected_takes
+
+        # mocking the two functions to check only the logic of get_valid_piece_moves
```
Comment (excellenteam-scaleup) [inline game_tests.py]: Good documentation. Implement it throughout your code!

**Review by excellenteam-scaleup — CHANGES_REQUESTED** [review]
Good job!
1. Missing most needed documentation
2. Assertions can be more indicative.

File: game_tests.py (hunk 1 above)
Comment (hamzash10) [reply game_tests.py]: documented all tests

File: game_tests.py (hunk 2 above)
Comment (hamzash10) [reply game_tests.py]: done

File: game_tests.py (hunk 3 above)
Comment (hamzash10) [reply game_tests.py]: sure, done documenting each function in great detail

File: game_tests.py (hunk 4 above)
Comment (hamzash10) [reply game_tests.py]: done

File: game_tests.py (hunk 5 above)
Comment (hamzash10) [reply game_tests.py]: done

**Review by excellenteam-scaleup — APPROVED** [review]

**Pull request merged**